### PR TITLE
Fix CI for Swift Package Manager

### PR DESCRIPTION
**Pull request by Patrick-Kladek**
Creating a PR here in hopes the CI will run.

This PR was done to finalize the work from #24. As the original Author is not responding we continue the official discussion here.

### CODE REVIEW

File: MapboxCoreNavigationTests/Heading.swift [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	
+	open override var headingAccuracy: CLLocationDirection {
+		get {
+			return _accuracy
+		}
+		set {
+			_accuracy = newValue
+		}
+	}
+}
```
Comment (StephanPartzsch) [inline MapboxCoreNavigationTests/Heading.swift:35]: Please stick to the established indentation style throughout the codebase in all files. Consistent indentation improves readability and ensures professionalism.

File: MapboxNavigation/Constants.swift [hunk 2]
```diff
@@ -16,6 +16,14 @@ public let MBRouteLineWidthByZoomLevel: [Int: NSExpression] = [
...
 ]
 
+public let MLNRouteLineWidthByZoomLevel: [Int: Double] = [
+	10: 8,
+	13: 9,
+	16: 11,
+	19: 22,
+	22: 28
+]
+
```
Comment (StephanPartzsch) [inline MapboxNavigation/Constants.swift:26]: Please stick to the established indentation style.

```suggestion
public let MLNRouteLineWidthByZoomLevel: [Int: Double] = [
    10: 8,
    13: 9,
    16: 11,
    19: 22,
    22: 28
]

```

File: MapboxNavigation/CarPlayNavigationViewController.swift [hunk 3]
```diff
@@ -197,7 +197,7 @@ public class CarPlayNavigationViewController: UIViewController, MLNMapViewDelega
         let location = notification.userInfo![RouteControllerNotificationUserInfoKey.locationKey] as! CLLocation
         
         // Update the user puck
-        let camera = MLNMapCamera(lookingAtCenter: location.coordinate, fromDistance: 120, pitch: 60, heading: location.course)
+		let camera = MLNMapCamera(lookingAtCenter: location.coordinate, acrossDistance: 120, pitch: 60, heading: location.course)
```
Comment (StephanPartzsch) [inline MapboxNavigation/CarPlayNavigationViewController.swift:200]:
```suggestion
	let camera = MLNMapCamera(lookingAtCenter: location.coordinate, acrossDistance: 120, pitch: 60, heading: location.course)
```

File: MapboxCoreNavigationTests/MD5Tests.swift [hunk 4]
```diff
@@ -49,7 +49,7 @@ class Tests: XCTestCase {
     }
     
     func testNaughtyStrings() {
-        let path = Bundle(for: Tests.self).path(forResource: "md5_crazy_strings", ofType: "txt")!
+		let path = Bundle.module.path(forResource: "md5_crazy_strings", ofType: "txt")!
```
Comment (StephanPartzsch) [inline MapboxCoreNavigationTests/MD5Tests.swift:52]:
```suggestion
	let path = Bundle.module.path(forResource: "md5_crazy_strings", ofType: "txt")!
```

File: MapboxCoreNavigationTests/Heading.swift [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import CoreLocation
+
+class Heading: CLHeading {
+	
+	private var _heading: CLLocationDirection
```
Comment (StephanPartzsch) [inline MapboxCoreNavigationTests/Heading.swift:5]: Please apply the naming style of the codebase (without the underscore).

File: MapboxNavigation/NavigationMapView.swift [hunk 6]
```diff
@@ -625,7 +625,10 @@ open class NavigationMapView: MLNMapView, UIGestureRecognizerDelegate {
                 arrow.minimumZoomLevel = minimumZoomLevel
                 arrow.lineCap = NSExpression(forConstantValue: "butt")
                 arrow.lineJoin = NSExpression(forConstantValue: "round")
-                arrow.lineWidth = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($zoomLevel, 'linear', nil, %@)", MBRouteLineWidthByZoomLevel.multiplied(by: 0.70))
+				arrow.lineWidth = NSExpression(forMLNInterpolating: .zoomLevelVariable,
+											   curveType: .linear,
+											   parameters: nil,
+											   stops: NSExpression(forConstantValue: MLNRouteLineWidthByZoomLevel.multiplied(by: 0.7)))
```
Comment (StephanPartzsch) [inline MapboxNavigation/NavigationMapView.swift:631]: Please stick to the established indentation style.

```suggestion
                arrow.lineWidth = NSExpression(
                    forMLNInterpolating: .zoomLevelVariable,
                    curveType: .linear,
                    parameters: nil,
                    stops: NSExpression(
                        forConstantValue: MLNRouteLineWidthByZoomLevel.multiplied(by: 0.7)
                    )
                )
```

File: MapboxNavigation/NavigationMapView.swift [hunk 7]
```diff
@@ -639,8 +642,11 @@ open class NavigationMapView: MLNMapView, UIGestureRecognizerDelegate {
                 arrowStroke.minimumZoomLevel = arrow.minimumZoomLevel
                 arrowStroke.lineCap = arrow.lineCap
                 arrowStroke.lineJoin = arrow.lineJoin
-                arrowStroke.lineWidth = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($zoomLevel, 'linear', nil, %@)", MBRouteLineWidthByZoomLevel.multiplied(by: 0.80))
-                arrowStroke.lineColor = NSExpression(forConstantValue: maneuverArrowStrokeColor)
+				arrow.lineWidth = NSExpression(forMLNInterpolating: .zoomLevelVariable,
```
Comment (StephanPartzsch) [inline MapboxNavigation/NavigationMapView.swift:645]: Same: indention

File: MapboxNavigation/NavigationMapView.swift [hunk 8]
```diff
@@ -666,7 +672,10 @@ open class NavigationMapView: MLNMapView, UIGestureRecognizerDelegate {
                 arrowSymbolLayer.iconColor = NSExpression(forConstantValue: maneuverArrowColor)
                 arrowSymbolLayer.iconRotationAlignment = NSExpression(forConstantValue: "map")
                 arrowSymbolLayer.iconRotation = NSExpression(forConstantValue: shaftDirection as NSNumber)
-                arrowSymbolLayer.iconScale = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($zoomLevel, 'linear', nil, %@)", MBRouteLineWidthByZoomLevel.multiplied(by: 0.12))
+				arrowSymbolLayer.iconScale = NSExpression(forMLNInterpolating: .zoomLevelVariable,
```
Comment (StephanPartzsch) [inline MapboxNavigation/NavigationMapView.swift:675]: Same: indention

File: MapboxNavigation/NavigationMapView.swift [hunk 9]
```diff
@@ -675,7 +684,10 @@ open class NavigationMapView: MLNMapView, UIGestureRecognizerDelegate {
                 arrowSymbolLayerCasing.iconColor = NSExpression(forConstantValue: maneuverArrowStrokeColor)
                 arrowSymbolLayerCasing.iconRotationAlignment = arrowSymbolLayer.iconRotationAlignment
                 arrowSymbolLayerCasing.iconRotation = arrowSymbolLayer.iconRotation
-                arrowSymbolLayerCasing.iconScale = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($zoomLevel, 'linear', nil, %@)", MBRouteLineWidthByZoomLevel.multiplied(by: 0.14))
+				arrowSymbolLayerCasing.iconScale = NSExpression(forMLNInterpolating: .zoomLevelVariable,
```
Comment (StephanPartzsch) [inline MapboxNavigation/NavigationMapView.swift:687]: Same: indention

File: MapboxNavigation/NavigationMapView.swift [hunk 10]
```diff
@@ -927,8 +939,10 @@ open class NavigationMapView: MLNMapView, UIGestureRecognizerDelegate {
     
     func routeStyleLayer(identifier: String, source: MLNSource) -> MLNStyleLayer {
         let line = MLNLineStyleLayer(identifier: identifier, source: source)
-        line.lineWidth = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($zoomLevel, 'linear', nil, %@)", MBRouteLineWidthByZoomLevel)
-        
+		line.lineWidth = NSExpression(forMLNInterpolating: .zoomLevelVariable,
```
Comment (StephanPartzsch) [inline MapboxNavigation/NavigationMapView.swift:942]: Same: indention

File: MapboxNavigation/NavigationMapView.swift [hunk 11]
```diff
@@ -949,7 +963,10 @@ open class NavigationMapView: MLNMapView, UIGestureRecognizerDelegate {
         let lineCasing = MLNLineStyleLayer(identifier: identifier, source: source)
         
         // Take the default line width and make it wider for the casing
-        lineCasing.lineWidth = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($zoomLevel, 'linear', nil, %@)", MBRouteLineWidthByZoomLevel.multiplied(by: 1.5))
+		lineCasing.lineWidth = NSExpression(forMLNInterpolating: .zoomLevelVariable,
```
Comment (StephanPartzsch) [inline MapboxNavigation/NavigationMapView.swift:966]: Same: indention

File: MapboxNavigationTests/Sources/Support/Constants.swift [hunk 12]
```diff
@@ -12,8 +12,8 @@ struct ShieldImage {
 }
 
 extension ShieldImage {
-    static let i280 = ShieldImage(image: UIImage(named: "i-280", in: Bundle(for: InstructionsBannerViewIntegrationTests.self), compatibleWith: nil)!,
+	static let i280 = ShieldImage(image: UIImage(named: "i-280", in: .module, compatibleWith: nil)!,
```
Comment (StephanPartzsch) [inline MapboxNavigationTests/Sources/Support/Constants.swift:15]:
```suggestion
    static let i280 = ShieldImage(image: UIImage(named: "i-280", in: .module, compatibleWith: nil)!,
```

File: MapboxNavigationTests/Sources/Support/Constants.swift [hunk 13]
```diff
@@ -12,8 +12,8 @@ struct ShieldImage {
...
                                   url: URL(string: "https://s3.amazonaws.com/mapbox/shields/v3/i-280@3x.png")!)
-    static let us101 = ShieldImage(image: UIImage(named: "us-101", in: Bundle(for: InstructionsBannerViewIntegrationTests.self), compatibleWith: nil)!,
+	static let us101 = ShieldImage(image: UIImage(named: "us-101", in: .module, compatibleWith: nil)!,
```
Comment (StephanPartzsch) [inline MapboxNavigationTests/Sources/Support/Constants.swift:17]:
```suggestion
    static let us101 = ShieldImage(image: UIImage(named: "us-101", in: .module, compatibleWith: nil)!,
```

File: MapboxNavigationTests/Sources/Support/Fixture.swift [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+internal extension Fixture {
+    
+    class func downloadRouteFixture(coordinates: [CLLocationCoordinate2D], fileName: String, completion: @escaping () -> Void) {
+        let accessToken = "<# Mapbox Access Token #>"
+        let directions = Directions(accessToken: accessToken)
+        
+        let options = RouteOptions(coordinates: coordinates, profileIdentifier: .automobileAvoidingTraffic)
+        options.includesSteps = true
+        options.routeShapeResolution = .full
+        let filePath = (NSTemporaryDirectory() as NSString).appendingPathComponent(fileName)
```
Comment (StephanPartzsch) [inline MapboxNavigationTests/Sources/Support/Fixture.swift:16]: Why do you need the cast? `NSTemporaryDirectory()` returns a `String`...

File: MapboxNavigationTests/Sources/Support/Fixture.swift [hunk 15]
```diff
@@ -0,0 +1,41 @@
+import XCTest
+import MapboxCoreNavigation
+import Foundation
+import MapboxDirections
+import CoreLocation
+
+internal extension Fixture {
```
Comment (StephanPartzsch) [inline MapboxNavigationTests/Sources/Support/Fixture.swift:7]: Please stick to the established indentation style.

File: MapboxNavigationTests/Sources/Tests/DataCacheTests.swift [hunk 16]
```diff
@@ -27,9 +27,9 @@ class DataCacheTests: XCTestCase {
 
     var exampleData: Data? {
```
Comment (StephanPartzsch) [inline MapboxNavigationTests/Sources/Tests/DataCacheTests.swift:28]: Please stick to the established indentation style.

File: MapboxNavigationTests/Sources/Tests/DummyURLSessionDataTask.swift [hunk 17]
```diff
@@ -0,0 +1,11 @@
+import Foundation
+
+class DummyURLSessionDataTask: URLSessionDataTask {
```
Comment (StephanPartzsch) [inline MapboxNavigationTests/Sources/Tests/DummyURLSessionDataTask.swift:3]: Is this used?
If so, please stick to the established indentation style and remove the empty comments.

File: MapboxNavigationTests/Sources/Tests/NavigationViewControllerTests.swift [hunk 18]
```diff
@@ -1,14 +1,16 @@
...
 @testable import MapboxNavigation
 
-let response = Fixture.JSONFromFileNamed(name: "route-with-instructions")
-let otherResponse = Fixture.JSONFromFileNamed(name: "route-for-lane-testing")
+let response = Fixture.JSONFromFileNamed(name: "route-with-instructions", bundle: .module)
+let otherResponse = Fixture.JSONFromFileNamed(name: "route-for-lane-testing", bundle: .module)
 
 class NavigationViewControllerTests: XCTestCase {
     
+	let fakeDirections = Directions(accessToken: "abc", host: "ab")
```
Comment (StephanPartzsch) [inline MapboxNavigationTests/Sources/Tests/NavigationViewControllerTests.swift:13]:
```suggestion
    let fakeDirections = Directions(accessToken: "abc", host: "ab")
```

**Review by StephanPartzsch — COMMENTED** [review]
Looks great, thanks for the work :)

**Comment by Patrick-Kladek** [comment]
@StephanPartzsch thank you for the review. 

I've noticed a lot of code style issues and wanted to ask if you use a code formatter? I started using [SwiftFormat](https://github.com/nicklockwood/SwiftFormat), configured it to reformat on commit automatically and it worked flawlessly for over 2 months now.

Would you accept this change, if I add it? I would let it configure itself from the current code style at first, and then tweak the settings if needed.

**Comment by StephanPartzsch** [comment]
@Patrick-Kladek 

At the moment there is no formatter configuration in the codebase, but it is a good idea to add one. This helps us having a consistent code style. 
Please use the indention level as it is used in the majority of files of this codebase.

**Comment by Patrick-Kladek** [comment]
I've also noticed there is a `.swiftlint` file in the repo, however, we are currently not using it. 

As adding these tools will introduce a huge amount of changes, I'll create a new PR from this branch, add SwiftFormat, also enable SwiftLint, and fix all the errors/warnings found by the tools. Once we can ensure a consistent code style we can go ahead and finalize this PR.

**Comment by boldtrn** [comment]
Yes I agree, let's do reformatting of the whole codebase in a separate PR. IMHO anything that we can automate and follows more or less the default XCode formatting is fine by me. Let's do one big reformatting PR and push this through to avoid too many conflicts.

**Comment by StephanPartzsch** [comment]
Having two separate PR is fine. 
We can finalize this one and add SwiftLint and swiftformat in a later PR as it will change the indention issues of this PR anyways. Or you can do it by hand for now @Patrick-Kladek . The number of issues wasn't that huge ;)

File: MapboxCoreNavigationTests/Heading.swift (hunk 1 above)
Comment (Patrick-Kladek) [reply MapboxCoreNavigationTests/Heading.swift:35]: Sure, if you tell me which indentation you mean, looking through the codebase I see different styles. I don't mind fixing it all to be consistent, however we should agree on one style first.

For example MLNMapView uses the same style
Bildschirmfoto 2024-04-11 um 11 12 21

File: MapboxCoreNavigationTests/Heading.swift (hunk 5 above)
Comment (Patrick-Kladek) [reply MapboxCoreNavigationTests/Heading.swift:5]: In this case, it is needed as this variable would clash with the super class. It is a workaround so we can mock this class for testing.

File: MapboxNavigationTests/Sources/Tests/DummyURLSessionDataTask.swift (hunk 17 above)
Comment (Patrick-Kladek) [reply MapboxNavigationTests/Sources/Tests/DummyURLSessionDataTask.swift:3]: yes this is used

**Comment by Patrick-Kladek** [comment]
Hi everyone, I've prepared another PR https://github.com/Patrick-Kladek/maplibre-navigation-ios/pull/2 to ensure consistent code style by auto formatting. Can you please take a first look at the style so we can discuss styling there further.

I'm not able to create a PR in repository as its base branch is in my Repo, I can only do that after this PR is merged.

**Review by StephanPartzsch — APPROVED** [review]

**Comment by StephanPartzsch** [comment]
LGTM

File: MapboxCoreNavigationTests/Heading.swift (hunk 1 above)
Comment (StephanPartzsch) [reply MapboxCoreNavigationTests/Heading.swift]: The issue is that there is no convention ;)
But this will be fixed with the formatter.

File: MapboxNavigationTests/Sources/Support/Fixture.swift (hunk 14 above)
Comment (StephanPartzsch) [reply MapboxNavigationTests/Sources/Support/Fixture.swift:16]: you are right

**Comment by Patrick-Kladek** [comment]
@boldtrn can you please review it as well and merge it

**Comment by boldtrn** [comment]
I think it is very unfortunate that we are loosing the examples with this change, maybe we can setup a second repo with the examples or something like this as a followup?

File: MapboxNavigation/CarPlayNavigationViewController.swift [hunk 19]
```diff
@@ -197,7 +197,7 @@ public class CarPlayNavigationViewController: UIViewController, MLNMapViewDelega
         let location = notification.userInfo![RouteControllerNotificationUserInfoKey.locationKey] as! CLLocation
         
         // Update the user puck
-        let camera = MLNMapCamera(lookingAtCenter: location.coordinate, fromDistance: 120, pitch: 60, heading: location.course)
+        let camera = MLNMapCamera(lookingAtCenter: location.coordinate, acrossDistance: 120, pitch: 60, heading: location.course)
```
Comment (boldtrn) [inline MapboxNavigation/CarPlayNavigationViewController.swift:200]: While I don't mind this change, it is a bit weird, that this is here. Is this intentional? Why is this changed? This seems to be unrelated to SPM?

File: MapboxNavigation/NavigationMapView.swift [hunk 20]
```diff
@@ -625,7 +625,10 @@ open class NavigationMapView: MLNMapView, UIGestureRecognizerDelegate {
                 arrow.minimumZoomLevel = minimumZoomLevel
                 arrow.lineCap = NSExpression(forConstantValue: "butt")
                 arrow.lineJoin = NSExpression(forConstantValue: "round")
-                arrow.lineWidth = NSExpression(format: "mgl_interpolate:withCurveType:parameters:stops:($zoomLevel, 'linear', nil, %@)", MBRouteLineWidthByZoomLevel.multiplied(by: 0.70))
+                arrow.lineWidth = NSExpression(forMLNInterpolating: .zoomLevelVariable,
+                                               curveType: .linear,
+                                               parameters: nil,
+                                               stops: NSExpression(forConstantValue: MLNRouteLineWidthByZoomLevel.multiplied(by: 0.7)))
```
Comment (boldtrn) [inline MapboxNavigation/NavigationMapView.swift:631]: Ah ok, this is about https://github.com/maplibre/maplibre-navigation-ios/issues/6?

If it works, I am fine with it, but in general I think we should try to limit PRs to one issue/topic as this is hidden between the SPM changes.

**Review by boldtrn — APPROVED** [review]
Thanks a lot for working on this PR and the effort that went into it, also from @MaartenZonneveld  👍 

I am not a fan of the hidden changes here and there. It's nothing major and if you made sure they work, I am fine with merging this as it is. I guess there might be some other issues hidden that we only find in integration testing later on, so let's merge this and fix stuff that might be wrong further down the line.

File: MapboxNavigation/CarPlayNavigationViewController.swift (hunk 19 above)
Comment (Patrick-Kladek) [reply MapboxNavigation/CarPlayNavigationViewController.swift:200]: This is the warning I got and thats why I changed it:

```
'init(lookingAtCenter:fromDistance:pitch:heading:)' is deprecated: Use -cameraLookingAtCenterCoordinate:acrossDistance:pitch:heading: or -cameraLookingAtCenterCoordinate:altitude:pitch:heading:.
```

Bildschirmfoto 2024-04-12 um 13 31 01

File: MapboxNavigation/NavigationMapView.swift (hunk 20 above)
Comment (Patrick-Kladek) [reply MapboxNavigation/NavigationMapView.swift:631]: Yes this is related to the errors logged to console. I don't feel comfortable releasing an app to our customers when I see errors. I could verify after the changes these errors are not printed anymore and navigation seems to work fine

**Comment by Patrick-Kladek** [comment]
> I am not a fan of the hidden changes here and there.

I rather fix small issues right away instead of opening up a discussion, and then spending 90% of time managing a one line change, that then drags on for days.

Retroactively speaking this could be the wrong approach here, so thanks for catching that. I'll try to limit the scope of my changes per PR from now on. 

Do you want me to revert the NSExpression change and submit it in a new PR?
